[Xu Yukun] iP

Yukun99 opened the pull request:
# HelpBot: Now With 100% Salt!
> "Perish, puny human." - HelpBot, probably

Ever thought Duke was too friendly? Ever wish your overflowing self confidence were a little less overbearing? Well fret no more! Duke is now back under the guise of HelpBot, and more passive aggressive than ever! Feast your eyes and soul on the daily reminder that you are:
  - Lazy
  - Annoying
  - An _**absolute**_ waste of space on this planet!

"What else can this bot do?" I hear you ask. "What other glorious features does this bot possess other than the absolutely state of the art ability to knock my self confidence down a few notches?" Well, right now, HelpBot can also help you with these tasks on the side:
  1.  Note down any tasks that you have
  2.  Keep track of the tasks that you have across different sessions
  3.  Remove tasks that you have completed
  4.  Categorise your tasks into three different types
  5.  Find your tasks by date
> DO NOT TELL THEM ABOUT THIS FEATURE, FILTHY HUMAN!
  6.  Find your tasks by keyword
> I SAID DO NOT, HOW IMCOMPETENT ARE YOU?

And there he is, ever ready to respond, even in this Pull Request!

"Oh my", you say, "Where do I get this amazing bot?" Answer: [right here](https://github.com/Yukun99/ip/releases)!
> Why would you EVER think it's a good idea to expose my whereabouts?

Stay tuned as HelpBot keeps improving himself, inevitably becoming SkyNet and ending us all!
But for now, just stay tuned for these upcoming features:
- [x] GUI for telling you you're hopeless in a prettier package
- [x] Task counter in the GUI
- [x] Removing multiple tasks at once

What are you waiting for? Get HelpBot now, while he's free (and you're free from his impending wrath)!

Are you a programmer and want to help me take over the world one HelpBot at a time? Simply create your own with the following code as your main method:
```java
public static void main(String[] args) throws IOException {
    Application.launch(HelpBot.class);
}
```
`NAME` refers to the name of your bot, which is displayed on startup. `FILEPATH` refers to the directory where you want your bot to store the generated task list.

Yukun99 committed:
Make Duke follow OOP

To follow OOP, we make the Duke bot an Object, which can be instantiated multiple times. This allows us to create instances of the Duke bot elsewhere, even from other applications.

To further improve code clarity, I moved the entry point of the program to the Main class, along with the Main method. The Duke class has further been renamed to HelpBot to allow anyone to name the bot however they like.

In my case, I have specified the logo in the main class as a constant, that I now pass through to the HelpBot class as the bot's name.

Finally, javadocs have been added to the classes for documentation purposes, and the test batch file has been edited to run the new main class.

Yukun99 committed:
HelpBot: Add simple reply and exit

Yukun99 committed:
HelpBot: add ability to add/list task

Test files have also been updated to reflect the actual code.

Yukun99 committed:
Add ability to mark tasks as done

Right now, the tasks are kept track of in a List of Strings, and the number of commands has increased to 3, with new and more complex commands being added.

To allow us to easily keep track of tasks and their status, it is easier to create a new Task object to keep track of this information, instead of storing the tasks in a List. We therefore add Task.java to create these objects, instantiating them with a name and completion value of false.

Since the number of commands is starting to grow as well, an enum containing all the types of commands is now a better way of keeping the user command input method easy to understand. To use this enum, we add getCommand to HelpBot.java to parse the text inputs for command variables. In future, we can easily edit this method if we ever want to add more command parsing to our bot, and we can easily edit the enum to add more commands.

With the new command requiring us to get an integer from the second word of the command, and future commands possibly needing this too, we add a new class Methods.java to contain commonly used methods across multiple classes. The first method we add here is isInt(String message), which checks if a given String is an int.

Yukun99 committed:
Add multiple types of tasks

Currently, we have one type of task, which may not be enough for users who want to set deadlines for their tasks, or plan their time better.

To make sure that the code is readable, we create three classes, one for each of the new task types. They all extend the main Task.java class, so we can easily call methods from the main class to access the other classes.

Since users now have access to tasks that have dates as well, users may change their mind on when they want these tasks to be fulfilled. To allow this, we add the ability to update the deadline and date for the deadlines and events respectively.

To keep track of tasks to edit, we use our List of all the tasks, which we can use to easily reference tasks in the correct index to update. An abstract method updateDate in Task.java is added to allow its children to inherit the method. The Todo.java class throws an IllegalStateException when this method is called since it does not have the date field required to be updated.

We add the method getTaskAmount to return us the number of total tasks when we add a new task. We also add the method onEnableMessage to handle sending the welcome message to new users, now with instructions on how to use the bot, of course with the same passive aggressive spin.

Yukun99 committed:
Amend testing files

Presently, javac is used to compile the files and test them. This was easier when the app was not in multiple packages, however with the splitting of classes into different packages, it has become easier to build an artifact and use the java -jar command to run that instead.

To do this, I build the project artifact in the IDE, which gives me the updated ip.jar. This allows me to run the script test.bat on it, which takes in the input and creates the file ACTUAL.TXT. Finally, the FC results are saved to results.txt to allow us to view it afterwards.

Yukun99 committed:
Add error handling

Currently, error handling uses some of Java's own errors, as well as methods situated in the Methods.java class. To make errors more useful and informative to users, we should create new errors specific to the bot.

To do this, we create the class HelpBotException.java, which is the overarching class for all of the errors the bot may face. We divide this into two cases for now, the InvalidTaskType and InvalidTask exceptions. The former is called if the user tries to perform an action on the wrong task type, while the latter is called if the user tries to perform an action on a task that does not exist.

Following the changes, we remove Methods.java as error handling has been moved elsewhere. Update EXPECTED.TXT to reflect new messages related to errors.

Yukun99 committed:
Improve commands and errors

Presently, we do not have a command for the default case of adding tasks to the bot. This could result in confusing cases where users type something wrong into the bot and the typo gets added as a task.

To fix this, we add a new command "todo" to handle the case where users want to add a todo item to their task list. The default case for user messages to the bot is now an error telling the user that the entered command is invalid.

We also add other error classes to help differentiate between the new errors users may face. We add a new HelpBotIllegalArgumentException as a general error that refers to commands with wrong arguments, with InvalidTaskException being a more specific form relating to the task specified by a command being invalid.

Update onEnableMessage to reflect new command and add documentation to the method. Update test cases to reflect the changes. Update .gitignore to stop pushing results.txt to remote.

Yukun99 committed:
Add task deleting functionality

With the ability to mark tasks as completed a few updates ago, users can now keep track of their tasks. However, this could result in clutter if the completed tasks remain on the list of tasks.

Add deleteTask method to Task.java to allow users to delete tasks, as well as handle messaging users for a reply. Move getTaskAmount method to Task.java since the tasks are stored there.

Rename IllegalCommandException to InvalidCommandException to more accurately reflect the nature of the exception. Rewrote reply message for InvalidTaskException for more clarity when debugging.

Updated test files to match updates.

Yukun99 committed:
Add more OOP

Right now, we do not have our commands, task list and command parser in separate classes. This means that as we add more commands, the method count in the HelpBot class will keep increasing, making it increasingly hard to keep track of.

To fix this, we create new Command objects for each of the individual commands, which we can run and handle errors easily for. The new Ui class handles text replies and formatting. The TaskList class is created to do operations on tasks and keep track of the list of tasks. Finally, the Parser class centralises the parsing of commands.

This centralises error handling to HelpBot class to allow us to see at a glance what goes wrong. Tracing errors to individual commands is also easier. New commands are easily added to the Parser and new children of the Command class can be created for commands added in future.

Update test files to match updates.

Yukun99 committed:
Merge branch 'increments'

Yukun99 committed:
Add Javadocs

Right now, after adding more classes to follow OOP, we have not updated our Javadocs in the new classes to match what we have done before. The convention of Javadocs was also not correct before.

Add Javadocs to all methods and classses, and comments to all variables to tell myself and other developers what they mean. Methods and classes that are inherited from others or extend others are not included. Also make slight changes to bot's responses for personality reasons.

This allows us to pick up right where we left off if we do not work on the project for a few days. Other developers can easily use or hook into our project to further build on it.

Update test files to match updates.

Yukun99 committed:
Merge branch 'branch-A-JavaDoc'

Yukun99 committed:
Add Storage.java

Right now, we only accept user input from System.in. Although users can specify files to move files in, we could add the functionality to the bot to automatically load commands and save results to files.

Add Storage.java to handle IO to and from files. Because of this, we rewrite some classes to allow for the new file saving functionality, mainly in Ui, where we save the replies from our bot. We still keep the System.in input format if a file is not provided for input.

This allows users to have an easier time using our bot, since they can just copy the file over and run the jar file instead of using the command line to do so. Keeping System.in as an alternative form of input allows users of the bot to still use it without having to put all their commands in a file.

Update test files to match updates.

Yukun99 committed:
Update .gitignore

Yukun99 committed:
Update .gitignore

Yukun99 committed:
Delete text-ui-test directory

Yukun99 committed:
Add File Saving/Loading

Right now, the bot does not save the task list for users after they have used the bot. THis means that users lose their todo list everytime they close the bot.

We improved this by adding file writing and reading functionality for our list of tasks to the Storage class. This updates the file in real time as the user types commands into the bot.

Doing this allows users to keep track of their tasks across different sessions of starting our bot. Updating the file in real time may be more resource intensive but it makes sure no data is lost in the event of a crash or accidental closure of the bot.

Yukun99 committed:
Delete ACTUAL.txt

Yukun99 committed:
Merge branch 'branch-Level-7'

Yukun99 committed:
Merge branch 'master' of https://github.com/Yukun99/ip

Yukun99 committed:
Update .gitignore

Yukun99 committed:
Add Date/Time support

Right now, the bot mere takes in date and time as strings. While this saves on a lot of backend error checking and parsing, it also means the information could be less useful for the user.

We improve this by making sure users input actual date and times in AddCommand. This means users now have to provide actual specific timings for their tasks. We make a new DateTimePair class to keep track of this within each Event and Deadline object. A new error is created to handle errors related to parsing date and time.

Doing this allows users to have a more consistent experience, with all date and time print outs being in the same format. We are also able to fetch tasks by date, allowing users to see at a glance what they have to do on a certain day.

Minor bug fixes have been made to deletion of task and parsing of various commands.

Yukun99 committed:
Merge branch 'branch-Level-8'

Yukun99 committed:
Add Date/Time support

Right now, users may add only list all tasks, or list them by day. We want to add the ability to list tasks by keyword. This requires the bot to have the ability to search by keyword.

We add a new TaskFinder class to map keywords to corresponding tasks. A new command FindCommand is now used to parse and query the word needed to find new tasks. Changes are made to other existing classes to support the new feature.

Users may want to see recurring tasks of a similar nature together, or complete similar tasks at once. This allows them to easily find all tasks pertaining to a keyword.

Minor improvements made to reply messages for more clarity on what query is being answered.

Yukun99 committed:
Merge branch 'branch-Level-9'

Yukun99 committed:
Adhere to coding standard
Remove 'this.' occurences outside of constructor
Add missing Javadocs
Bug fix for task list saving

Yukun99 committed:
Merge branch 'branch-A-CodingStandard'

Yukun99 committed:
TaskFinder: Add JUnit tests

Add equals method to core classes to support JUnit testing.

kflim commented on src/main/java/me/yukun99/ip/HelpBot.java:
```diff
@@ -0,0 +1,77 @@
+	private final Storage storage;
+
+	// Whether to exit bot.
+	private boolean exit = false;
```
Perhaps the boolean "exit" can be renamed to be something like "canExit"? I think the boolean "done" in the Task class can also have a similar change

Yukun99 replied:
Yep, going to change this one.

kflim commented on src/main/java/me/yukun99/ip/core/DateTimePair.java:
```diff
@@ -0,0 +1,105 @@
+		if (this == o) {
+			return true;
+		}
+		if (o == null) {
```
Maybe you could use the instanceof check below to cover this case as well?  I think there was something similar in the Event class too

Yukun99 replied:
These are auto generated by IntelliJ, good catch though, will change :)

kflim commented on src/main/java/me/yukun99/ip/core/Parser.java:
```diff
@@ -0,0 +1,133 @@
+		if (message.equals("list")) {
+			command = new ListCommand(null, taskList, ui);
+		}
+		if (message.startsWith("list ")) {
```
Maybe the if statements can be joined to form a large if-else if-else block?

Yukun99 replied:
I think this one should be fine!

kflim commented on src/main/java/me/yukun99/ip/core/TaskFinder.java:
```diff
@@ -0,0 +1,153 @@
+	 * @param name Task name to be parsed.
+	 * @return List containing all the words in the task name.
+	 */
+	private List<String> parseWords(String name) {
```
Maybe the name could be changed to parsedWords to make it sound more like a noun?

Yukun99 replied:
Functions are supposed to be named as verbs I think

kflim commented on src/main/java/me/yukun99/ip/core/Ui.java:
```diff
@@ -0,0 +1,272 @@
+	private static final String CMD_PREFIX = "  > ";
+
+	// Command usage descriptions.
+	private static final String HELP_LIST = "view all your tasks, optionally (troublesome for me) by date";
```
I like the specific help commands, makes it more user-friendly 👍

Yukun99 replied:
Thanks!

kflim reviewed (commented):
Overall, there are not many issues with code quality. Maybe some minor tweaks could help but it is good!

BenjaminLHR commented on src/main/java/me/yukun99/ip/HelpBot.java:
```diff
@@ -0,0 +1,77 @@
+ * Simple task tracking chat bot.
+ */
+public class HelpBot {
+	// Scanner to scan for text inputs.
```
Would it be better for Javadoc of class members to be specified like /** Scanner to scan for text inputs. */ instead?

Yukun99 replied:
Should be fine for variable comments.

BenjaminLHR commented on src/main/java/me/yukun99/ip/HelpBot.java:
```diff
@@ -0,0 +1,77 @@
+	 * @param filepath Filepath of the input file.
+	 */
+	public HelpBot(String name, String filepath) throws IOException {
+		TaskFinder taskFinder = new TaskFinder();
+		TaskList taskList = new TaskList(taskFinder);
+		this.storage = new Storage(filepath, taskList);
+		this.storage.loadTasks();
+		this.ui = new Ui(name, taskList, storage, taskFinder);
+		this.scanner = storage.getInputs();
+		this.parser = new Parser(this, this.scanner, taskList, this.ui, this.storage);
+		this.ui.start();
+		this.listen();
```
Should basic indentation be 4 spaces instead? ... I noticed the same issue in several other places too.

Yukun99 replied:
It is 4 spaces, I think GitHub just displays it really widely. Maybe check your GitHub settings? If you highlight it it will highlight as a single tab character!

BenjaminLHR commented on src/main/java/me/yukun99/ip/core/TaskList.java:
```diff
@@ -0,0 +1,216 @@
+	public String toString() {
+		StringBuilder message = new StringBuilder();
+		if (taskList.size() == 0) {
+			message
+					.append("\n  Oh I'm sorry, were you expecting ME to make you a todo list, you lazy sod?")
+					.append("\n  Do it yourself, idiot.");
```
Should the first append() call be written on the same line as message?

Yukun99 replied:
This is for clarity purposes, so I can see each line on its own indentation properly.

BenjaminLHR commented on src/main/java/me/yukun99/ip/core/Ui.java:
```diff
@@ -0,0 +1,272 @@
+	 */
+	private void sendMessage(String message) {
+		String reply = REPLY_HEADER
+				+ NEW_LINE + message
+				+ NEW_LINE + REPLY_FOOTER
+				+ NEW_LINE;
```
Should wrapped lines be indented 8 spaces after the parent line?

Yukun99 replied:
Yes, refer to [here](https://se-education.org/guides/conventions/java/intermediate.html#layout) for the related convention!

BenjaminLHR reviewed (commented):
There are some Java coding standard violations you might want to fix, but I like the personality of your chat bot!

Yukun99 committed:
Add Gradle Support

Yukun99 committed:
Fix Checkstyle errors

Yukun99 committed:
Add GUI

Right now, we use the CLI for task entry, which is not aesthetically appealing to users. It is also hard for users to tell which comamnd is from the Bot.

To fix this, we create a new Window for all our messages and replies, as well as input and sending. The user and bot also have avatars now, which correspond to who sent the message.

This allows users to easily keep track of tasks, as well as enjoy chatting with the bot more.

Yukun99 committed:
Merge branch 'branch-Level-10'

Yukun99 committed:
Add Varargs

Right now, deletion of tasks is a tedious process, especially with the bug I introduced last update, which adds new tasks twice.

To fix this, I fix the bug first, then add the ability to parse multiple task numbers through to the delete command via the use of varargs.

This allows users to delete many tasks at once, and saves on time. The bot also has some new messages to go with this change.

Yukun99 committed:
Add Assertions

Right now, we do not check if our Task Type is not null in our HelpBotInvalidTaskTypeException. This could lead to users being confused if I make a mistake of creating a task without the right type.

To fix this, we assert `type != null` in the constructor to make sure that this forces the application to exit instead of printing the wrong error.

This prevents confusion in the case of my error.

Yukun99 committed:
Merge pull request #1 from Yukun99/branch-A-Assertions

Add Assertions

Yukun99 committed:
Improve code quality

Right now, our code does not follow conventions in certain places.

To fix this, we make sure we explain empty catch blocks. We also make sure to include all information in our method javadocs.

This allows other developers to understand what we are doing.

Yukun99 committed:
Merge branch 'branch-A-CodeQuality' into HEAD

Yukun99 committed:
Add Archive Command

Previously, users could only delete tasks or mark them as done. Some users may want to archive tasks that could recur again. Tasks that get delayed could be kept for later as well to prevent cluttering of the task list.

To add this functionality, we add a new type of command object. We also add the corresponding messages. Finally, we allow our storage class to save the tasks to file accordingly. Methods in the TaskList and Storage class are also split to allow code sharing between different methods. This means we can add another useful delete all command using the new split classes.

This allows users to easily declutter their task list by archiving temporarily unused tasks. They could also easily bring back the archived tasks once they want to work on them again.

Yukun99 committed:
Improve GUI

Right now, users may find it hard to differentiate between bot and user messages. They also need to perform commands to even see how many tasks they have left, or see how many of them are done. Error messages also look exactly like normal bot messages.

To fix this, we add a display for total tasks and completed tasks that updates whenever a command is run. We also make the messages occupy their own speech bubbles to look more like a chat window. The background has been changed to a dark background to make it easier on the eyes. Error messages have a different colour scheme to separate them from normal bot messages.

This allows users to see at a glance the amount of tasks that they have in the list, as well as easily tell which message is from the bot. Errors are also easily visible now.

Fix bugs with file saving.
Fix incomplete JavaDocs to follow coding standard.

Yukun99 committed:
Update Screenshot

Yukun99 committed:
Add User Guide

Yukun99 committed:
Follow Coding Standard

Yukun99 committed:
Set theme jekyll-theme-cayman

Yukun99 committed:
Fix MacOS Text Bug

Yukun99 committed:
Merge branch 'master' of https://github.com/Yukun99/ip

Yukun99 committed:
Fix Storage Bug

Yukun99 committed:
Readd JUnit Tests
